### PR TITLE
Feature/backend/oauth-get oauthToken, refreshOauthToken and get RSI remote data with Refit.

### PR DESCRIPTION
**Pull request by peggy-gov-bc-ca**
# Description

This PR includes the following proposed change(s):

- TCVP-182 create OAuthHandler to get Token from bc gov oauth server
- TCVP-184 User Refit to call RSI rest endpoint to get ticket info

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Tested locally and unit tests

## Does the change impact or break the Docker build?

- [ ] Yes
- [X] No

If Yes: Has Docker been updated accordingly?

- [ ] Yes
- [X] No

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            _memoryCache = memoryCache;
+            _logger = logger;
+        }
+
+        public async Task<Token> GetTokenAsync(CancellationToken cancellationToken)
+        {
+            var token = _memoryCache.Get<Token>(Keys.OAUTH_TOKEN_KEY);
+            if (token == null)
+            {
+                _logger.LogInformation("no token in memory, refresh token.");
```
Comment (pbolduc) [inline src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs:31]: Would Debug level make more sense?  The log message could be more clear, ie, "Cached token not found, requesting a new token"

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                return await RefreshTokenAsync(cancellationToken);
+            }
+            return token;
+        }
+
+        private async Task<Token> RefreshTokenAsync(CancellationToken cancellationToken)
+        {
+            Token token = await _oAuthCient.GetRefreshToken(cancellationToken);
+            _logger.LogInformation("get new token and save it to memory");
```
Comment (pbolduc) [inline src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs:40]: Add additional information to the log message. Which key will the token be stored, when will the token cache expire?

**Review by pbolduc — CHANGES_REQUESTED** [review]
Here are some additional feedback after reviewing the code in Visual Studio.

File: .gitignore [hunk 3]
```diff
@@ -40,6 +41,11 @@ TestResult.xml
 #RESHARPER
 .DotSettings
 
+# .NET Core
+project.lock.json
+project.fragment.lock.json
+artifacts/
+**/Properties/launchSettings.json
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply .gitignore:48]: ok, I remove it. and clean the lauchSettings.json

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                        : await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                    _logger.LogError($"get oauth token failed. {responseData}"); 
+                    throw new OAuthException(
+                        "The HTTP status code of the response was not expected (" + (int)response.StatusCode + ").",
+                        (int)response.StatusCode, responseData,
+                        response.Headers.ToDictionary(x => x.Key, x => x.Value), null);
+                }
+
+                _logger.LogInformation("get oauth token successfully");
+                var stream = await response.Content.ReadAsStreamAsync();
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs:62]: sure, added

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+                return await JsonSerializer.DeserializeAsync<Token>(stream);
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs:64]: sure

[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs:64]: fixed

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                {"resource", _oAuthOptions.ResourceUrl},
+                {"client_id", _oAuthOptions.ClientId},
+                {"client_secret", _oAuthOptions.Secret},
+                {"scope", "openid"},
+                {"response_mode", "form_post"},
+                {"grant_type", "client_credentials"}
+            };
+
+            var content = new FormUrlEncodedContent(data);
+            using (var request = new HttpRequestMessage(HttpMethod.Post, _oAuthOptions.OAuthUrl) { Content = content })
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs:44]: fixed

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        public OAuthClient(HttpClient httpClient, IOptionsMonitor<OAuthOptions> oAuthOptions, ILogger<OAuthClient> logger)
+        {
+            _httpClient = httpClient;
+            _oAuthOptions = oAuthOptions.CurrentValue;
+            _logger = logger;
+        }
+
+        public async Task<Token> GetRefreshToken(CancellationToken cancellationToken)
+        {
+            _httpClient.DefaultRequestHeaders.Add("return-client-request-id", "true");
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/OAuthClient.cs:30]: fixed

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace DisputeApi.Web.Auth
+{
+    public interface ITokenService    {
+        Task<Token> GetTokenAsync(CancellationToken cancellationToken);
+    }
+
+    public class TokenService : ITokenService
+    {
+        private IOAuthClient _oAuthCient;
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs:15]: fixed

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs (hunk 1 above)
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs:31]: fixed

File: src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs (hunk 2 above)
Comment (peggy-gov-bc-ca) [reply src/backend/TrafficCourtsApi/DisputeApi.Web/Auth/TokenService.cs:40]: fixed

**Pull request merged**